src/nvim/window.c

```diff
@@ -937,10 +937,21 @@ void ui_ext_win_viewport(win_T *wp)
       // interact with incomplete final line? Diff filler lines?
       botline = wp->w_buffer->b_ml.ml_line_count;
     }
+    int scroll_delta = 0;
+    if (wp->w_viewport_last_topline > line_count) {
+      scroll_delta += wp->w_viewport_last_topline - line_count;
+      wp->w_viewport_last_topline = line_count;
+    }
+    if (wp->w_topline > wp->w_viewport_last_topline) {
+      scroll_delta += plines_m_win(wp, wp->w_viewport_last_topline, wp->w_topline - 1);
+    } else if (wp->w_topline < wp->w_viewport_last_topline) {
+      scroll_delta -= plines_m_win(wp, wp->w_topline, wp->w_viewport_last_topline - 1);
```

Comment:
Can scroll_delta be negative? If so, is that covered by the tests? If not, should we add an assert or some other guard around here?

Reply:
I wrote all of this quite some while ago so I don't quite remember all the details. But yes, `scroll_delta` can be negative, and this isn't covered by tests yet. Gonna add some, but I will have to look into how to actually define tests here again :)

Reply:
I did actually already have a test for this. See my other comment.